kv.go

```diff
@@ -18,22 +18,36 @@ type KVService struct {
 }
 
 // KVSetOption is an option passed to Set() operation.
-type KVSetOption func(*model.PluginKVSetOptions)
+type KVSetOption func(*model.PluginKVSetOptions) error
 
 // SetAtomic guarantees the write will occur only when the current value of matches the given old
 // value. A client is expected to read the old value first, then pass it back to ensure the value
 // has not since been modified.
 func SetAtomic(oldValue interface{}) KVSetOption {
-	return func(o *model.PluginKVSetOptions) {
+	return func(o *model.PluginKVSetOptions) error {
 		o.Atomic = true
-		o.OldValue = oldValue
+		if oldValue != nil {
+			oldValueBytes, isOldValueInBytes := oldValue.([]byte)
+			if isOldValueInBytes {
```

Comment:
This /seems/ like the wrong place to do this -- why not just check in the main call instead?

Reply:
If I move the code to the main call, I would have to store `oldValue` somewhere, but `PluginKVSetOptions` only has a `OldValue []byte`.

Reply:
Could we use an unexported struct embedding and extending `model.PluginKVSetOptions` instead? I'd love to keep the business logic here as closely coupled together as possible, vs. splitting it between here and the main call.

Reply:
I applied your suggestion in c570432d10ce8044b632ad56caf3b1af4922044b.

While I see the point keeping the logic close together, with this change the coupling between the downstream opts and the opts struct used here is lost. Maybe this is a good thing, because new options downstream are not exposed to the public immediately, but it make the code harder to read and understand.

I'm fine merging this PR with the changes, but want to keep the discussion on how close the downstream code could be coupled going.